clippy_lints/src/loops.rs

```diff
@@ -1953,6 +1986,54 @@ fn check_for_single_element_loop<'tcx>(
     }
 }
 
+/// Check if a for loop loops over `Option`s or `Result`s and contains only
+/// a `if let Some` or `if let Ok` expression.
+fn check_manual_flatten<'tcx>(cx: &LateContext<'tcx>, pat: &'tcx Pat<'_>, arg: &'tcx Expr<'_>, body: &'tcx Expr<'_>) {
+    if_chain! {
+        // Ensure the `if let` statement is the only expression in the for-loop
+        if let ExprKind::Block(ref block, _) = body.kind;
+        if block.stmts.is_empty();
+        if let Some(inner_expr) = block.expr;
+        if let ExprKind::Match(_ , _, MatchSource::IfLetDesugar{ contains_else_clause }) = inner_expr.kind;
+        if !contains_else_clause;
+        // Ensure the `if let` statement is for `Option` or `Result`
+        let pat_is_option = is_type_diagnostic_item(cx, cx.typeck_results().pat_ty(pat).peel_refs(), sym::option_type);
+        let pat_is_result = is_type_diagnostic_item(cx, cx.typeck_results().pat_ty(pat).peel_refs(), sym::result_type);
+        if pat_is_option || pat_is_result;
+
+        then {
+            // Prepare the error message
+            let pat_type = if pat_is_option {
+                "Option"
+            } else {
+                "Result"
+            };
+            let msg = format!("for loop over `{}`s with a single `if let` expression.", pat_type);
```

Comment:
For the message and help message, let me suggest... (feel free to disagree 😃)

> Unnecessary `if let` since only the `Some` variant of the iterator element is used
> ...
> help: try `x.iter().flatten()`